src/arch/riscv/rvdbg013.c

```diff
@@ -0,0 +1,1772 @@
+			rvdbg_dmi_read_pure(dmi, DMI_REG_ABSTRACTDATA0, NULL);
+		} else if (len > 8) {
+			rvdbg_dmi_read_pure(dmi, DMI_REG_ABSTRACTDATA0, &data);
+		} else {
+			rvdbg_dmi_read_nop(dmi, &data);
+		}
+		if (dmi->error) {
+			DEBUG_WARN("Read at len %d failed\n", len);
+			return;
+		}
+		size_t chunk = MIN(len, 4);
+		memcpy(dest, &data, chunk);
+		dest += chunk;
+		len -= chunk;
+	}
+	if (dmi->error) {
+		DEBUG_WARN("Abstract read failed at addr %p\n", dest);
+		return;
+	}
+	rvdbg_dmi_read(dmi, DMI_REG_ABSTRACT_CS, &data);
+	if (ABSTRACTCS_GET_CMDERR(data)) {
+		DEBUG_WARN("mem_read_abstract failure\n");
+		dmi->error = true;
+	}
+}
+static void rvdbg_mem_write_abstract(target *t, target_addr dest, const void* src, size_t len)
+{
+	RVDBGv013_DMI_t *dmi = t->priv;
+	if (!dest) {
+		DEBUG_WARN("rvdbg_mem_write_abstract invalid buffer\n");
+		dmi->error = true;
+		return;
+	}
+	if (!len)
+		return;
+	const uint32_t *w;
+	/* Align to word */
+	if (dest & 3) {
+		const uint8_t *p = src;
+		size_t count = MIN(4 - (dest & 3), len);
+		len -= count;
+		uint32_t cmd = ABSTRACTCMD_TYPE_ACCESS_MEMORY | ABSTRACTCMD_AAMSIZE_8bit | ABSTRACTCMD_WRITE;
+		while (count && !dmi->error) {
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACTDATA1, dest);
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACTDATA0, *p);
+			dest ++;
+			p++;
+			count--;
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACT_CMD, cmd);
+		}
+		w = (const uint32_t *) p;
+		if (!len)
+			return;
+	} else {
+		w = src;
+	}
+	if (dmi->error)
+		return;
+	if (len > 3) {
+		uint32_t cmd =  ABSTRACTCMD_TYPE_ACCESS_MEMORY | ABSTRACTCMD_AAMSIZE_32bit | ABSTRACTCMD_WRITE;
+		rvdbg_dmi_write(dmi, DMI_REG_ABSTRACTDATA1, dest);
+		rvdbg_dmi_write(dmi, DMI_REG_ABSTRACTDATA0, *w);
+		len -= 4;
+		dest += 4;
+		w++;
+		if (len > 3 ) {
+			cmd |= ABSTRACTCMD_AAMPOSTINCREMENT;
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACT_CMD, cmd);
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACT_AUTOEXEC, ABSTRACTAUTO_AUTOEXECDATA);
+		} else {
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACT_CMD, cmd);
+		}
+		while ((len > 3) && !dmi->error) {
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACTDATA0, *w);
+			w++;
+			len -= 4;
+			dest += 4;
+			if (len < 4) /* Terminate writing words*/ {
+				rvdbg_dmi_write(dmi, DMI_REG_ABSTRACT_AUTOEXEC, 0);
+				break;
+			}
+		}
+	}
+	if (len && !dmi->error) {
+		uint8_t *p = (uint8_t *)w;
+		uint32_t cmd =  ABSTRACTCMD_TYPE_ACCESS_MEMORY | ABSTRACTCMD_AAMSIZE_8bit | ABSTRACTCMD_WRITE;
+		while (len && !dmi->error) {
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACTDATA1, dest);
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACTDATA0, *p);
+			dest++;
+			p++;
+			len--;
+			rvdbg_dmi_write(dmi, DMI_REG_ABSTRACT_CMD, cmd);
+		}
+	}
+	uint32_t data;
+	rvdbg_dmi_read(dmi, DMI_REG_ABSTRACT_CS, &data);
+	if (ABSTRACTCS_GET_CMDERR(data)) {
+		DEBUG_WARN("mem_read_abstract failure\n");
+		dmi->error = true;
+	}
+}
+
+static void rvdbg_mem_read_systembus(target *t,  void* dest, target_addr address, size_t len)
```

Comment:
maybe the different mem read & write implementations should multiplex at the byte & word r/w functions like rvdbg_mem_read_systembus, and generalize the repeated align to word + copy the rest in words logic?